### PR TITLE
added more bundle tests

### PR DESCRIPTION
**Pull request by technosophos**
Lo and Behold! By writing silly tests, I found a bug! It's almost like there's a reason we do this testing thing!

Added a whole bunch of tests, and started to learn more about Spectral.

### CODE REVIEW

File: src/cnab.rs [hunk 1]
```diff
@@ -18,7 +18,7 @@ pub struct Bundle {
     /// The list of configurable credentials.
     ///
     /// Credentials are injected into the bundle's invocation image at startup time.
-    pub credentials: Option<Vec<Credential>>,
+    pub credentials: Option<HashMap<String, Credential>>,
```
Comment (radu-matei) [inline src/cnab.rs]: Wow, this testing thing really is something!

**Review by radu-matei — APPROVED** [review]
LGTM.

Also, I really like the blocks for testing various parts of the bundle -- all memory assigned in the block is freed, right? I like it!

**Comment by technosophos** [comment]
I'm going to try this one again after rebasing with the String -> str change.

**Comment by simonferquel** [comment]
Also, would be nice to put some of the basic tests in function/struct documentation as in https://doc.rust-lang.org/rustdoc/documentation-tests.html.

That makes tests being a source of documentation (which is one of the awesome rustlang feature, ensuring documentation is aligned with code changes)

**Comment by technosophos** [comment]
@radu_matei can I merge this one? I think both of us are now working off of this instead of master.

**Comment by radu-matei** [comment]
Sure, I already LGTM’d it, and I’ve branched off this, so merge away.

File: src/cnab.rs [hunk 2]
```diff
@@ -11,134 +11,158 @@ use std::path::Path;
...
+    pub fn from_string(json_data: &str) -> Result<Bundle, BundleParseError> {
         let res: Bundle = serde_json::from_str(json_data)?;
         Ok(res)
     }
 
-    pub fn from_file(file_path: &str) -> Result<Bundle, serde_json::Error> {
-        let file = File::open(Path::new(&file_path)).expect("file not found");
-        let res: Bundle = serde_json::from_reader(file)?;
+    pub fn from_file(file_path: &str) -> Result<Bundle, BundleParseError> {
+        //let file = File::open(Path::new(&file_path)).expect("file not found");
```
Comment (silvin-lubecki) [inline src/cnab.rs]: nit: there is some commented code, it should be removed?

File: src/tests.rs [hunk 3]
```diff
@@ -206,3 +364,11 @@ fn test_bundle_deserialize() {
     assert_that(&bun.maintainers.unwrap().len()).is_equal_to(&1);
     assert_that(&bun.custom.unwrap().len()).is_equal_to(&2);
 }
+
+
+// Check that a missing file results in an error (not a panic)
+#[test]
+fn test_bundle_from_file_not_found() {
+    let bun = Bundle::from_file("no/such/file.json");
+    assert_that(&bun.is_err()).is_true();
```
Comment (radu-matei) [inline src/tests.rs:373]: Well, this panicked:

```
thread 'tests::test_bundle_from_file_not_found' panicked at 'file not found: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/libcore/result.rs:997:5
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

Comment (technosophos) [reply src/tests.rs:373]: @radu-matei can you please retest this?

I rebased, and if I run `cargo test`, all seems to be passing now.

**Review by radu-matei — APPROVED** [review]
Tested again, LGTM.

**Pull request merged**